File: 06/06.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <time.h>

#define FILE_SIZE 4096

char* read_data(const char* file_name, size_t file_size) {
    FILE* fp = fopen(file_name, "r");
    if (fp == NULL) return NULL;

    char* data = malloc(sizeof(char) * file_size);
    if (data == NULL) {
        fclose(fp);
        return NULL;
    }

    if (fread(data, sizeof(char), file_size, fp) != file_size - 1) {
        free(data);
        fclose(fp);
        return NULL;
    }
    fclose(fp);

    return data;
}

int solve(char* data, size_t data_len, size_t n_distinct) {
    for (int i = n_distinct; i < data_len; i++) {
        for (int j = i - n_distinct; j < i; j++) {
            for (int k = i - n_distinct; k < i; k++) {
                if (j != k && data[j] == data[k]) {
                    goto not_header;
                }
            }
        }
        // no duplicate in last n_distinct characeters
        return i;

        not_header:
        // duplicate found, just continue
    }
    return -1;
}

unsigned long timestamp_nano() {
    struct timespec time;
    timespec_get(&time, TIME_UTC);
    return time.tv_nsec;
}

int main() {
    unsigned long start = timestamp_nano();

    char* data = read_data("06/input.txt", FILE_SIZE);

    int p1 = solve(data, FILE_SIZE - 1, 4);
    printf("part1: %d\n", p1);

    int p2 = solve(data, FILE_SIZE - 1, 14);
    printf("part2: %d\n", p2);

    free(data);

    unsigned long end = timestamp_nano();
    printf("time: %.1fµs\n", (end - start) / 1000.0);

    exit(EXIT_SUCCESS);
}

File: 06/06.rs
use std::fs;
use std::time::Instant;

fn main() {
    let start = Instant::now();

    let data: String = fs::read_to_string("06/input.txt").expect("File not found");

    let p1 = solve(&data, 4, 0);
    println!("part1: {}", p1);    
    println!("part1: {}", solve(&data, 14, (p1 - 4) as usize));    

    println!("time: {:?}", start.elapsed());
}

fn solve(data: &String, n_distinct: usize, offset: usize) -> i32 {
    let bytes = data.as_bytes();
    'outer: for i in (offset + n_distinct)..bytes.len() {
        // for-loop go brrrrr
        for j in i - n_distinct..i {
            for k in i - n_distinct..i {
                if j != k && bytes[j] == bytes[k] {
                    continue 'outer;
                }
            }
        }
        return i as i32;
    }
    return -1;
}
